lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -308,48 +330,36 @@ impl ManagerInner {
-        // to obtain them, create a mapping for the target page to verify that
-        // the target PFN is valid. This mapping will be used later if this is
-        // the first overlay for the target PFN.
+        // Moving an overlay consists of applying it to its new location and
+        // then removing it from its old one. This is only legal if the target
+        // PFN is valid and there is no other overlay of this kind associated
+        // with that PFN.
+        //
+        // Checking these conditions up front allows this function to attempt to
+        // remove the existing overlay (checking for errors) and then assert
+        // that the overlay can be applied in its new position. This is simpler
+        // than removing the old overlay, failing to add the new one, and then
+        // having to remember how to add the old overlay back in its previous
+        // position.
         let mut to_mapping = to_pfn.map(&memctx)?;
+        if let Some(to_set) = self.overlays.get(&to_pfn) {
+            if to_set.contains_kind(kind) {
+                return Err(OverlayError::KindInUse(to_pfn.into(), kind));
+            }
+        }
 
-        // Get the page contents to transfer, but don't actually change the
-        // tables yet, since applying the new overlay might still fail if there
-        // is an existing overlay of this kind at the target PFN.
-        //
-        // It's possible to do this check now, modify the tables in-place, and
-        // then assert that adding the new overlay succeeded, but this sequence
-        // allows this function to reuse the existing overlay removal code. This
-        // is unlikely to be a bottleneck in practice because guests change
-        // their overlay locations so infrequently (and any guest that spends
-        // all its time moving overlays around has, in some ways, asked for
-        // it...).
-        let contents = if from_set.active == kind {
-            let mut from_mapping = from_pfn
-                .map(&memctx)
-                .expect("active overlay PFNs can be mapped");
-            let mut contents = OverlayContents::default();
-            from_mapping.read_page(&mut contents);
-            contents
-        } else if let Some(pending_contents) = from_set.pending.get(&kind) {
-            pending_contents.clone()
-        } else {
-            return Err(OverlayError::NoOverlayOfKind(from_pfn.into(), kind));
-        };
+        // If the overlay exists in its old location, then it can definitely be
+        // moved to its new one, so it's now safe to try removing the old
+        // overlay.
+        let mut contents = Some(MaybeUninit::uninit());
+        self.remove_overlay(from_pfn, kind, acc_mem, contents.as_mut())?;
+
+        // Safety: `remove_overlay` guarantees that the output contents will be
+        // initialized on success.
+        let contents = unsafe { contents.unwrap().assume_init() };
```

Comment:
Hm, I don't think I see why all this `MaybeUninit` business is necessary. Naively, it _looks_ to me like `remove_overlay` should be able to just return an `OverlayContents` which we could use here?

Reply:
Ah, I see, the intent is that whether or not the `Option<&mut MaybeUninit>` is passed in dictates whether or not we return the previous contents, so that we can avoid reading the old overlay page contents if we're just removing it rather than moving it. Hm. That does make sense although I have to admit I'm not terribly enamored with it...

Reply:
I reimplemented this in f09684e with slightly different semantics (caller passes a flag, routine returns an `Option`). Turned out a bit nicer, I think--now the caller will panic if the callee forgot to return contents instead of invoking UB, and it's not really any less economical from a lines-of-code perspective than the MaybeUninit version.